helix-core/src/main/java/org/apache/helix/task/TaskDriver.java

```diff
@@ -699,6 +798,73 @@ private void addTaskToJobConfig(String workflowName, String jobName, TaskConfig
     throw new TimeoutException("An unexpected issue happened while task being added to the job!");
   }
 
+  /**
+   * A helper method which deletes an existing task from the job config and verifies if task is
+   * deleted from the context by the controller.
+   * @param workflowName
+   * @param jobName
+   * @param taskID
+   * @param endTime
+   * @throws InterruptedException
+   * @throws TimeoutException
+   */
+  private void deleteTaskFromJobConfig(String workflowName, String jobName, String taskID,
```

Comment:
I haven't reviewed these code completely. But my guess is that this method is different from addTaskToJobConfig() for only the updater part. All the rest are the same. Could you please merge?

Reply:
So they seem to be similar but they are different. Aside from the logs and updaters which are different, in one of them we add a task to config and then check if it is being added to context by the controller by polling. In another one, we delete from config and make sure it is being deleted by the controller from the context by polling. Anyways following your comment I did a few simplifications on add API. Thanks.

Reply:
Logs can be changed to "failed to update the task from the job...", etc.
The updater can be changed to be a parameter.

As for the additional check you mentioned, is it necessary? Since the updater is done, task modification has to be finished in the ZK. If the rebalancer runs slow, then the timeout will fail the operation even the task has been added/removed correctly.
Even we do want to wait, it is still doable by passing a verifier-style callback and then we can merge the private methods.
Or, alternatively, you can check the jobContext in the upper layer method after calls this generic task modification method. Note that since you are calling the accessor 2 times for the jobcontext and workflowcontext, there is no guarantee that the result is consistent anyway. So moving the jobcontext check outside will not introduce more issues.

Reply:
Ok. Yeah, we do want to check before returning the call. However, I think your comment makes sense. Let me try that and see how it goes.

Reply:
Done. Merged the update part in one method.